src/event/finalization.rs

```diff
@@ -0,0 +1,358 @@
+
+    #[cfg(test)]
+    fn count_finalizers(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// An event finalizer is the shared data required to handle tracking
+/// the status of an event, and updating the status of a batch with that
+/// when the event is dropped.
+#[derive(Debug)]
+pub struct EventFinalizer {
+    status: Atomic<EventStatus>,
+    batch: Arc<BatchNotifier>,
+}
+
+impl EventFinalizer {
+    /// Create a new event in a batch.
+    pub fn new(batch: Arc<BatchNotifier>) -> Self {
+        let status = Atomic::new(EventStatus::Dropped);
+        Self { status, batch }
+    }
+
+    /// Update this finalizer's status in place with the given `EventStatus`
+    pub fn update_status(&self, status: EventStatus) {
+        self.status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |old_status| {
+                Some(old_status.update(status))
+            })
+            .unwrap_or_else(|_| unreachable!());
+    }
+
+    /// Update the batch for this event with this finalizer's
+    /// status, and mark this event as no longer requiring update.
+    pub fn update_batch(&self) {
+        let status = self
+            .status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |_| {
+                Some(EventStatus::NoOp)
+            })
+            .unwrap_or_else(|_| unreachable!());
+        self.batch.update_status(status);
+    }
+}
+
+impl Drop for EventFinalizer {
+    fn drop(&mut self) {
+        self.update_batch();
+    }
+}
+
+/// A batch notifier contains the status of the current batch along with
+/// a one-shot notifier to send that status back to the source. It is
+/// shared among all events of a batch.
+#[derive(Debug)]
+pub struct BatchNotifier {
+    status: Atomic<BatchStatus>,
+    notifier: Option<oneshot::Sender<BatchStatus>>,
+}
+
+impl BatchNotifier {
+    /// Create a new `BatchNotifier` along with the receiver used to
+    /// await its finalization status.
+    pub fn new_with_receiver() -> (Arc<Self>, oneshot::Receiver<BatchStatus>) {
+        let (sender, receiver) = oneshot::channel();
+        let notifier = Self {
+            status: Atomic::new(BatchStatus::Delivered),
+            notifier: Some(sender),
+        };
+        (Arc::new(notifier), receiver)
+    }
+
+    /// Update this notifier's status from the status of a finalized event.
+    pub fn update_status(&self, status: EventStatus) {
+        self.status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |old_status| {
+                Some(old_status.update_from_event(status))
+            })
+            .unwrap_or_else(|_| unreachable!());
+    }
+
+    /// Send this notifier's status up to the source
+    pub fn send_status(&mut self) {
+        if let Some(notifier) = self.notifier.take() {
+            let status = self.status.load(Ordering::Relaxed);
+            if notifier.send(status).is_err() {
+                warn!(message = "Could not send batch acknowledgement notifier.");
```

Comment:
Could this happen if the source is dropped/has finished?

Reply:
This is the only way the error can happen actually. Ref: https://docs.rs/tokio/1.5.0/tokio/sync/oneshot/struct.Sender.html#method.send I didn't really want to leave this path without a warning, though, as it could indicate something else going wrong. Do you think it should be silent?

Reply:
Probably not silent, but rate limited. The case I'm worried about is when we are forcibly shutting down sources which will cause all of the outstanding batches to fail here which could be quite a number of them.

Reply:
Ah yes, this definitely needs a rate limit.

Reply:
Actually, given that shutdown is a normal event, we don't want to send a warning at all. I'll just drop the warning.

Reply:
I suppose that depends on the behavior of sources. Do/Should they wait for all of the batches to be confirmed before shutting down? If yes, then we can leave this warning.